### PR TITLE
Cleanup instance tracker.

### PR DESCRIPTION
**Pull request by jeschkies**
Summary:
The instance tracker would hold its state as part of the receive method.
With this change it will just be another field in the actor. We also cleanup
`InstancesBySpec` and `SpecInstances` a little.

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/InstanceTrackerActor.scala [hunk 1]
```diff
@@ -97,13 +101,19 @@ private[impl] class InstanceTrackerActor(
   override def receive: Receive = initializing
 
   private[this] def initializing: Receive = LoggingReceive.withLabel("initializing") {
-    case appTasks: InstanceTracker.InstancesBySpec =>
+    case initialInstances: InstanceTracker.InstancesBySpec =>
       logger.info("Task loading complete.")
 
+      instancesBySpec = initialInstances
+      counts = TaskCounts(initialInstances.allInstances, healthStatuses = Map.empty)
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/core/task/tracker/impl/InstanceTrackerActor.scala]: I guess we could make the task counts part if the `InstencesBySpec` and the also update the metrics there. What do you think?

Comment (ichernetsky) [reply src/main/scala/mesosphere/marathon/core/task/tracker/impl/InstanceTrackerActor.scala]: `the InstencesBySpec and the` — what comes after the second `the`?

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/InstanceTrackerActor.scala [hunk 2]
```diff
@@ -97,13 +101,19 @@ private[impl] class InstanceTrackerActor(
...
+
+      // this is run on any state change
```
Comment (ichernetsky) [inline src/main/scala/mesosphere/marathon/core/task/tracker/impl/InstanceTrackerActor.scala]: I believe this one should be removed.

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/InstancesLoaderImpl.scala [hunk 3]
```diff
@@ -25,12 +25,7 @@ private[tracker] class InstancesLoaderImpl(repo: InstanceRepository)(implicit va
       instances <- Future.sequence(names.map(repo.get)).map(_.flatten)
     } yield {
       log.info(s"Loaded ${instances.size} tasks")
-      val instancesByRunSpec = instances.groupBy(_.runSpecId)
-      val map = instancesByRunSpec.map {
-        case (runSpecId, specInstances) => runSpecId ->
-          InstanceTracker.SpecInstances.forInstances(runSpecId, specInstances)
```
Comment (alenkacz) [inline src/main/scala/mesosphere/marathon/core/task/tracker/impl/InstancesLoaderImpl.scala]: why did we have to it like this previously?

Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/core/task/tracker/impl/InstancesLoaderImpl.scala]: This was just code duplication that got in the base with the infamous `Team` merge for pods.

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/InstanceTrackerActor.scala (hunk 1 above)
Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/core/task/tracker/impl/InstanceTrackerActor.scala]: Ah, it should be `and then also update the metrics there`.

File: src/main/scala/mesosphere/marathon/core/task/tracker/InstanceTracker.scala [hunk 4]
```diff
@@ -78,27 +78,31 @@ object InstanceTracker {
...
+            val instancesById: Map[Instance.Id, Instance] = appInstances.map(instance => instance.instanceId -> instance)(collection.breakOut)
+            appId -> SpecInstances(instancesById)
+        }
     )
 
     def empty: InstancesBySpec = of(collection.immutable.Map.empty[PathId, InstanceTracker.SpecInstances])
   }
   /**
-    * Contains only the tasks of the app with the given app ID.
+    * Contains only the instances of a specific app.
```
Comment (zen-dog) [inline src/main/scala/mesosphere/marathon/core/task/tracker/InstanceTracker.scala]: nite: `s/specific app/specific RunSpec`

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/InstanceTrackerActor.scala [hunk 5]
```diff
@@ -97,13 +101,18 @@ private[impl] class InstanceTrackerActor(
   override def receive: Receive = initializing
 
   private[this] def initializing: Receive = LoggingReceive.withLabel("initializing") {
-    case appTasks: InstanceTracker.InstancesBySpec =>
+    case initialInstances: InstanceTracker.InstancesBySpec =>
       logger.info("Task loading complete.")
```
Comment (zen-dog) [inline src/main/scala/mesosphere/marathon/core/task/tracker/impl/InstanceTrackerActor.scala]: nit:`s/Task loading/Instances...`

File: src/main/scala/mesosphere/marathon/core/task/tracker/InstanceTracker.scala [hunk 6]
```diff
@@ -78,27 +78,31 @@ object InstanceTracker {
...
     *
-    * @param specId The id of the app.
-    * @param instanceMap The tasks of this app by task ID. FIXME: change keys to Task.TaskID
+    * @param instanceMap The instances of the app by their id.
     */
-  case class SpecInstances(specId: PathId, instanceMap: Map[Instance.Id, Instance] = Map.empty) {
+  case class SpecInstances(instanceMap: Map[Instance.Id, Instance] = Map.empty) {
```
[the remark this answers is not in the document]
Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/core/task/tracker/InstanceTracker.scala]: I was thinking about removing them. However, the have a few helper methods for construction. So I'd rather follow up with a removal.

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/InstanceTrackerActor.scala (hunk 1 above)
Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/core/task/tracker/impl/InstanceTrackerActor.scala]: Yes. However, if these classes are removed that would not be an option.

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/InstanceTrackerActor.scala (hunk 5 above)
Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/core/task/tracker/impl/InstanceTrackerActor.scala]: ✅

File: src/main/scala/mesosphere/marathon/core/task/tracker/InstanceTracker.scala (hunk 4 above)
Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/core/task/tracker/InstanceTracker.scala]: ✅

**Pull request merged**